apps/snitch_api/lib/snitch_api/products_context/products_context.ex

```diff
@@ -0,0 +1,137 @@
+defmodule SnitchApi.ProductsContext do
+  @moduledoc """
+  The JSON-API context.
+  """
+  alias Snitch.Repo
+  alias Snitch.Data.Schema.Product
+
+  import Ecto.Query, only: [from: 2, order_by: 2]
+
+  @doc """
+  List out the all products
+  """
+  def list_products(conn, params) do
+    query = define_query(params)
+    page = create_page(query, %{}, conn)
+    products = paginate_collection(query, params)
+    {products, page}
+  end
+
+  @doc """
+  Gives the product, with matched `slug` or raise an error :not_found
+  """
+  def product_by_slug!(slug) do
+    Product
+    |> Repo.get_by!(slug: slug)
+    |> Repo.preload(variants: [:images])
+  end
+
+  @doc """
+  Creates the page. The page comprises all the related pagination links
+  """
+  def create_page(query, params, conn) do
+    query
+    |> Repo.all()
+    |> gen_page_links(params, conn)
+  end
+
+  @doc """
+  Executes the given `query` applying the page in params.
+  """
+  def paginate_collection(query, params) do
+    {page_number, size} = extract_page_params(params)
+
+    query
+    |> paginate(page_number, size)
+    |> Repo.all()
+    |> Repo.preload(variants: [:images])
+  end
+
+  @doc """
+  Generates the pagination links like `prev` `self` `next` `last` using 
+  data-collection and params-page
+  """
+  def gen_page_links(collection, params, conn) do
+    {number, size} = extract_page_params(params)
+
+    JaSerializer.Builder.PaginationLinks.build(
+      %{
+        number: number,
+        size: size,
+        total: div(Enum.count(collection), size),
+        base_url: "http://" <> get_base_url(conn) <> "/api/v1/products"
```

Comment:
why `http://` only. Also this does not look like correct approved.

Reply:
Initially I thought `base_url: get_base_url(conn) <> "/api/v1/products"` For quick navigation I added `http`